### PR TITLE
[FEATURE][API] Add parts iterators to QgsGeometry 

### PR DESCRIPTION
**Pull request by nyalldawson**
This allows easy iteration over all the parts of a geometry,
regardless of the geometry's type. E.g.

     geometry = QgsGeometry.fromWkt( 'MultiPoint( 0 0, 1 1, 2 2)' )
     for part in geometry.parts():
       print(part.asWkt())

     geometry = QgsGeometry.fromWkt( 'LineString( 0 0, 10 10 )' )
     for part in geometry.parts():
       print(part.asWkt())

There are two iterators available. QgsGeometry.parts() gives
a non-const iterator, allowing the parts to be modified in place:

     geometry = QgsGeometry.fromWkt( 'MultiPoint( 0 0, 1 1, 2 2)' )
     for part in geometry.parts():
        part.transform(ct)

For a const iteration, calling .const_parts() gives a const
iterator, which cannot edit the parts but avoids a potentially expensive
QgsGeometry detach and clone

     geometry = QgsGeometry.fromWkt( 'MultiPoint( 0 0, 1 1, 2 2)' )
     for part in geometry.const_parts():
        print(part.x())

Question: should this be QgsGeometry.const_parts(), or QgsGeometry.constParts() ?

### CODE REVIEW

**Comment by nyalldawson** [comment]
I've renamed to "constParts" for consistency

**Comment by NathanW2** [comment]
Oh very nice!

File: python/core/auto_generated/geometry/qgsgeometry.sip.in [hunk 1]
```diff
@@ -353,11 +353,105 @@ Will return a negative value if a geometry is missing.
...
+   The iterator returns a copy of individual vertices, and accordingly geometries cannot be
+   modified using the iterator. See transformVertices() for a safe method to modify vertices "in-place".
+
+* Example:
+.. code-block:: python
+
+       # print the x and y coordinate for each vertex in a LineString
+       geometry = QgsGeometry.fromWkt( 'LineString( 0 0, 1 1, 2 2)' )
+       for v in geometry.vertices():
+          print(v.x(), v.y())
```
Comment (m-kuhn) [inline python/core/auto_generated/geometry/qgsgeometry.sip.in]: Standard python indentation is 4 ;)

File: src/core/geometry/qgsabstractgeometry.cpp [hunk 2]
```diff
@@ -245,6 +246,28 @@ void QgsAbstractGeometry::transformVertices( const std::function<QgsPoint( const
   // Ideally this would be pure virtual, but SIP has issues with that
 }
 
+QgsAbstractGeometry::part_iterator QgsAbstractGeometry::parts_end()
```
Comment (m-kuhn) [inline src/core/geometry/qgsabstractgeometry.cpp]: Don't we want these camelcase?

**Comment by m-kuhn** [comment]
Nice :+1:

File: src/core/geometry/qgsabstractgeometry.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/geometry/qgsabstractgeometry.cpp]: I've followed the vertex iterator format here -- which I believe was probably done in this way to follow the std iterator convention. So here I'd rather keep consistency with the vertex iterator which is already part of stable API.

Comment (m-kuhn) [reply src/core/geometry/qgsabstractgeometry.cpp]: Looking at Qt, they use this style whenever they have the text "Provided for STL compatibility." within the API docs. I could imagine to follow their lead. It's a good criteria to decide if it's acceptable in a given place.

I'm a tiny bit afraid that having these methods as precedence in the API in combination with a couple of pythonistas (or std c library devs) working on the code might lead to slowly shifting towards inconsistent style in the codebase.

Comment (nyalldawson) [reply src/core/geometry/qgsabstractgeometry.cpp]: I'm unsure - are you totally against this name, or do you want me to leave it but add the extra note to the dox?

Comment (m-kuhn) [reply src/core/geometry/qgsabstractgeometry.cpp]: Sorry for being unclear.

I'm in of consistency with the vertex iterator and keeping it as is.

And at the same time was off-topicly considering future standards for us to follow to not get trapped in "preference" discussions.
I don't think the extra note would make sense here, `parts_end` is rather _stl style_ than _stl compatibility_.

Comment (nyalldawson) [reply src/core/geometry/qgsabstractgeometry.cpp]: I've added a QGIS 4.0 TODO note to revisit this discussion.

**Pull request merged**